File: src/karabo/core/CameraFsm.cc
/*
 * $Id$
 *
 * Author: <burkhard.heisen@xfel.eu>
 *
 * Copyright (C) European XFEL GmbH Hamburg. All rights reserved.
 */

#include "CameraFsm.hh"
#include <karabo/xms/SlotElement.hh>

using namespace std;
using namespace karabo::util;
using namespace karabo::xms;
using namespace log4cpp;

namespace karabo {
    namespace core {


        void CameraFsm::expectedParameters(karabo::util::Schema& expected) {

            SLOT_ELEMENT(expected).key("slotEndError")
                    .displayedName("Reset")
                    .description("Resets the camera in case of an error")
                    .allowedStates("Error")
                    .commit();

            SLOT_ELEMENT(expected).key("slotAcquire")
                    .displayedName("Acquire")
                    .description("Instructs camera to go into acquisition state")
                    .allowedStates("AllOk.Ready")
                    .commit();

            SLOT_ELEMENT(expected).key("slotStop")
                    .displayedName("Stop")
                    .description("Instructs camera to stop current acquisition")
                    .allowedStates("AllOk.Acquisition")
                    .commit();

            SLOT_ELEMENT(expected).key("slotTrigger")
                    .displayedName("Trigger")
                    .description("Sends a software trigger to the camera")
                    .allowedStates("AllOk.Acquisition")
                    .commit();
        }


        void CameraFsm::configure(const karabo::util::Hash& input) {
            SLOT0(slotEndError)
            SLOT0(slotAcquire)
            SLOT0(slotStop)
            SLOT0(slotTrigger)
        }


        void CameraFsm::run() {
            startStateMachine();
            runEventLoop();
        }


        void CameraFsm::errorStateOnEntry() {

        }


        void CameraFsm::errorStateOnExit() {

        }


        void CameraFsm::hardwareSetupStateOnEntry() {

        }


        void CameraFsm::hardwareSetupStateOnExit() {

        }


        void CameraFsm::acquisitionStateOnEntry() {

        }


        void CameraFsm::acquisitionStateOnExit() {

        }


        void CameraFsm::readyStateOnEntry() {

        }


        void CameraFsm::readyStateOnExit() {

        }


        void CameraFsm::acquireAction() {

        }


        void CameraFsm::stopAction() {

        }


        void CameraFsm::triggerAction() {

        }
    }
}

File: src/pythonKarabo/karabo_package/camera_fsm.py
# To change this template, choose Tools | Templates
# and open the template in the editor.

__author__="andrea.parenti@xfel.eu"
__date__ ="August  7, 2013"

from karabo.decorators import KARABO_CLASSINFO
import karabo.base_fsm as base
from karabo.karathon import SLOT_ELEMENT
from karabo.fsm import *

@KARABO_CLASSINFO("CameraFsm", "1.0")
class CameraFsm(base.BaseFsm):
    
    @staticmethod
    def expectedParameters(expected):
        (
        SLOT_ELEMENT(expected).key("acquire")
        .displayedName("Acquire")
        .description("Instructs camera to go into acquisition state")
        .allowedStates("Ok.Ready")
        .commit()
        ,
        SLOT_ELEMENT(expected).key("trigger")
        .displayedName("Trigger")
        .description("Sends a software trigger to the camera")
        .allowedStates("Ok.Acquisition")
        .commit()
        ,
        SLOT_ELEMENT(expected).key("stop")
        .displayedName("Stop")
        .description("Instructs camera to stop current acquisition")
        .allowedStates("Ok.Acquisition")
        .commit()
        ,
        SLOT_ELEMENT(expected).key("reset")
        .displayedName("Reset")
        .description("Resets the camera in case of an error")
        .allowedStates("Error")
        .commit()
        )
    
    def __init__(self, configuration):
        super(CameraFsm, self).__init__(configuration)
        
        #**************************************************************
        #*                        Events                              *
        #**************************************************************
        KARABO_FSM_EVENT(self, 'ErrorFoundEvent', 'errorFound')
        KARABO_FSM_EVENT(self, 'ResetEvent',      'reset')
        KARABO_FSM_EVENT(self, 'AcquireEvent',    'acquire')
        KARABO_FSM_EVENT(self, 'StopEvent',       'stop')
        KARABO_FSM_EVENT(self, 'TriggerEvent',    'trigger')

        #**************************************************************
        #*                        States                              *
        #**************************************************************
        KARABO_FSM_STATE('Error', self.errorStateOnEntry, self.errorStateOnExit)
        KARABO_FSM_STATE('Initialization', self.initializationStateOnEntry, self.initializationStateOnExit)
        KARABO_FSM_STATE('Acquisition', self.acquisitionStateOnEntry, self.acquisitionStateOnExit)
        KARABO_FSM_STATE('Ready', self.readyStateOnEntry, self.readyStateOnExit)
        
        #**************************************************************
        #*                    Transition Actions                      *
        #**************************************************************
        #KARABO_FSM_NO_TRANSITION_ACTION(self.noStateTransition)    
        KARABO_FSM_ACTION('ErrorFoundAction', self.errorFoundAction, str, str)
        KARABO_FSM_ACTION('ResetAction', self.resetAction)
        KARABO_FSM_ACTION('AcquireAction', self.acquireAction)
        KARABO_FSM_ACTION('StopAction',  self.stopAction)
        KARABO_FSM_ACTION('TriggerAction',  self.triggerAction)
        
        #**************************************************************
        #*                       Ok State Machine                     *
        #**************************************************************
        okStt=[
        # Source-State   Event           Target-State   Action           Guard
        ('Ready',       'AcquireEvent', 'Acquisition', 'AcquireAction', 'none'),
        ('Acquisition', 'StopEvent',    'Ready',       'StopAction',    'none'),
        ('Acquisition', 'TriggerEvent', 'none',        'TriggerAction', 'none')
        ]
        #                        Name  Transition-Table  Initial-State
        KARABO_FSM_STATE_MACHINE('Ok', okStt, 'Ready')
        
        #**************************************************************
        #*                       Top Machine                          *
        #**************************************************************
        cameraStt=[
        # Source-State      Event        Target-State   Action              Guard
        ('Initialization', 'none',            'Ok',    'none',             'none'),
        ('Ok',             'ErrorFoundEvent', 'Error', 'ErrorFoundAction', 'none'),
        ('Error',          'ResetEvent',      'Ok',    'none',             'none')
        ]
        #                         Name      Transition-Table  Initial-State
        KARABO_FSM_STATE_MACHINE('CameraMachine', cameraStt, 'Initialization')
        self.fsm = KARABO_FSM_CREATE_MACHINE('CameraMachine')
        
    def getFsm(self):
        return self.fsm
    
    def initializationStateOnEntry(self):
        '''Actions executed on entry to 'Initialization' state
        '''
        
    def initializationStateOnExit(self):
        '''Actions executed on exit from 'Initialization' state
        '''
        
    def errorStateOnEntry(self):
        '''Actions executed on entry to 'Error' state
        '''
        
    def errorStateOnExit(self):
        '''Actions executed on exit from 'Error' state
        '''
        
    def acquisitionStateOnEntry(self):
        '''Actions executed on entry to 'Acquisition' state
        '''
        
    def acquisitionStateOnExit(self):
        '''Actions executed on exit from 'Acquisition' state
        '''
        
    def readyStateOnEntry(self):
        '''Actions executed on entry to 'Ready' state
        '''
        
    def readyStateOnExit(self):
        '''Actions executed on exit from 'Ready' state
        '''
        
    def acquireAction(self):
        '''Actions executed at 'acquire' event'''
        
    def stopAction(self):
        '''Actions executed at 'stop' event'''
        
    def triggerAction(self):
        '''Actions executed at 'trigger' event'''

    def resetAction(self):
        print("Reset action executed")
        
